File: pkg/clients/claude/parser.go
package claude

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"
	"time"

	"github.com/halfwhey/allagent/pkg/wrap"
)

// tailer tails a Claude Code JSONL session file and emits wrap.Events.
type tailer struct {
	path   string
	events chan wrap.Event
	cancel context.CancelFunc
}

func newTailer(path string) *tailer {
	ctx, cancel := context.WithCancel(context.Background())
	t := &tailer{
		path:   path,
		events: make(chan wrap.Event, 256),
		cancel: cancel,
	}
	go t.run(ctx)
	return t
}

func (t *tailer) events_ch() <-chan wrap.Event { return t.events }

func (t *tailer) close() {
	t.cancel()
}

func (t *tailer) run(ctx context.Context) {
	defer close(t.events)

	for {
		if _, err := os.Stat(t.path); err == nil {
			break
		}
		select {
		case <-ctx.Done():
			return
		case <-time.After(250 * time.Millisecond):
		}
	}
	var (
		offset  int64
		pending string
	)
	for {
		chunk, nextOffset, err := readAppended(t.path, offset)
		if err == nil && len(chunk) > 0 {
			offset = nextOffset
			pending += string(chunk)
			pending = t.emitCompleteLines(ctx, pending)
		} else if err == nil {
			offset = nextOffset
		}

		select {
		case <-ctx.Done():
			return
		case <-time.After(200 * time.Millisecond):
		}
	}
}

func (t *tailer) emitCompleteLines(ctx context.Context, pending string) string {
	for {
		idx := strings.IndexByte(pending, '\n')
		if idx < 0 {
			return pending
		}

		line := strings.TrimRight(pending[:idx], "\r")
		pending = pending[idx+1:]
		if line == "" {
			continue
		}
		for _, ev := range parseLineEvents(line) {
			select {
			case t.events <- ev:
			case <-ctx.Done():
				return pending
			}
		}
	}
}

func readAppended(path string, offset int64) ([]byte, int64, error) {
	fi, err := os.Stat(path)
	if err != nil {
		return nil, offset, err
	}
	if fi.Size() < offset {
		offset = 0
	}
	if fi.Size() == offset {
		return nil, offset, nil
	}

	f, err := os.Open(path)
	if err != nil {
		return nil, offset, err
	}
	defer f.Close()

	buf := make([]byte, fi.Size()-offset)
	n, err := f.ReadAt(buf, offset)
	if err != nil && err != io.EOF {
		return nil, offset, err
	}
	return buf[:n], offset + int64(n), nil
}

// -- JSON envelope types -----------------------------------------------------

type envelope struct {
	Type      string          `json:"type"`
	Subtype   string          `json:"subtype"`
	Timestamp string          `json:"timestamp"`
	Message   json.RawMessage `json:"message"`
	Content   string          `json:"content"`
	Data      json.RawMessage `json:"data"`
}

type message struct {
	Role    string          `json:"role"`
	Content json.RawMessage `json:"content"`
}

type contentBlock struct {
	Type      string          `json:"type"`
	Text      string          `json:"text"`
	Thinking  string          `json:"thinking"`
	ID        string          `json:"id"`
	Name      string          `json:"name"`
	Input     json.RawMessage `json:"input"`
	ToolUseID string          `json:"tool_use_id"`
	Content2  json.RawMessage `json:"content"`
}

// -- Parsing -----------------------------------------------------------------

func parseLineEvents(line string) []wrap.Event {
	var env envelope
	if err := json.Unmarshal([]byte(line), &env); err != nil {
		return nil
	}
	ts, _ := time.Parse(time.RFC3339Nano, env.Timestamp)

	switch env.Type {
	case "user":
		return parseMessageLineEvents(env, ts)
	case "system":
		if env.Content != "" {
			return []wrap.Event{{Kind: wrap.EventSystem, Time: ts, Text: env.Content, Role: "system"}}
		}
		return nil
	case "progress":
		if ev, ok := parseProgress(env, ts); ok {
			return []wrap.Event{ev}
		}
		return nil
	default:
		if len(env.Message) > 0 {
			return parseMessageLineEvents(env, ts)
		}
		return nil
	}
}

func parseMessageLineEvents(env envelope, ts time.Time) []wrap.Event {
	if len(env.Message) == 0 {
		return nil
	}
	var msg message
	if err := json.Unmarshal(env.Message, &msg); err != nil {
		return nil
	}
	if len(msg.Content) == 0 {
		return nil
	}

	// Try plain string content first.
	var str string
	if err := json.Unmarshal(msg.Content, &str); err == nil {
		kind := wrap.EventUserMessage
		if msg.Role == "assistant" {
			kind = wrap.EventAssistantText
		}
		return []wrap.Event{{Kind: kind, Time: ts, Text: str, Role: msg.Role}}
	}

	// Array of content blocks.
	var blocks []contentBlock
	if err := json.Unmarshal(msg.Content, &blocks); err != nil {
		return nil
	}

	var events []wrap.Event
	for _, b := range blocks {
		switch b.Type {
		case "text":
			kind := wrap.EventAssistantText
			if msg.Role == "user" {
				kind = wrap.EventUserMessage
			}
			events = append(events, wrap.Event{Kind: kind, Time: ts, Text: b.Text, Role: msg.Role})
		case "thinking":
			if b.Thinking != "" {
				events = append(events, wrap.Event{Kind: wrap.EventAssistantThink, Time: ts, Text: b.Thinking, Role: "assistant"})
			}
		case "tool_use":
			inp := "{}"
			if len(b.Input) > 0 {
				inp = string(b.Input)
			}
			events = append(events, wrap.Event{
				Kind:      wrap.EventToolCall,
				Time:      ts,
				Text:      fmt.Sprintf("%s(%s)", b.Name, truncate(inp, 80)),
				ToolName:  b.Name,
				ToolID:    b.ID,
				ToolInput: inp,
				Role:      "assistant",
			})
		case "tool_result":
			text := extractToolResultText(b)
			events = append(events, wrap.Event{
				Kind:   wrap.EventToolResult,
				Time:   ts,
				Text:   truncate(text, 200),
				ToolID: b.ToolUseID,
				Role:   "user",
			})
		}
	}
	return events
}

func parseProgress(env envelope, ts time.Time) (wrap.Event, bool) {
	if len(env.Data) == 0 {
		return wrap.Event{}, false
	}
	var d struct {
		Type  string `json:"type"`
		Query string `json:"query"`
	}
	if err := json.Unmarshal(env.Data, &d); err != nil {
		return wrap.Event{}, false
	}
	text := d.Type
	if d.Query != "" {
		text += ": " + d.Query
	}
	return wrap.Event{Kind: wrap.EventSystem, Time: ts, Text: text, Role: "system"}, true
}

func extractToolResultText(b contentBlock) string {
	if len(b.Content2) > 0 {
		var s string
		if json.Unmarshal(b.Content2, &s) == nil {
			return s
		}
		var blocks []struct {
			Type string `json:"type"`
			Text string `json:"text"`
		}
		if json.Unmarshal(b.Content2, &blocks) == nil {
			var parts []string
			for _, bb := range blocks {
				if bb.Text != "" {
					parts = append(parts, bb.Text)
				}
			}
			return strings.Join(parts, "\n")
		}
		return string(b.Content2)
	}
	return "(no output)"
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}

File: packages/claudraband-core/src/claude/parser.ts
import { stat, open } from "node:fs/promises";
import type { Event } from "../wrap/event";
import { EventKind, makeEvent } from "../wrap/event";

interface Envelope {
  type: string;
  subtype?: string;
  timestamp?: string;
  message?: unknown;
  content?: string;
  data?: unknown;
}

interface Message {
  role: string;
  content: unknown;
  stop_reason?: string | null;
}

interface ContentBlock {
  type: string;
  text?: string;
  thinking?: string;
  id?: string;
  name?: string;
  input?: unknown;
  tool_use_id?: string;
  content?: unknown;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

function truncate(s: string, n: number): string {
  if (s.length <= n) return s;
  return s.slice(0, n) + "...";
}

function extractToolResultText(b: ContentBlock): string {
  const raw = b.content;
  if (raw != null) {
    if (typeof raw === "string") return raw;
    if (Array.isArray(raw)) {
      const parts: string[] = [];
      for (const bb of raw as { type?: string; text?: string }[]) {
        if (bb.text) parts.push(bb.text);
      }
      if (parts.length > 0) return parts.join("\n");
    }
    return JSON.stringify(raw);
  }
  return "(no output)";
}

function parseProgress(env: Envelope, ts: Date): Event | null {
  if (!env.data || typeof env.data !== "object") return null;
  const d = env.data as { type?: string; query?: string };
  let text = d.type ?? "";
  if (d.query) text += ": " + d.query;
  return makeEvent({ kind: EventKind.System, time: ts, text, role: "system" });
}

function parseMessageLineEvents(env: Envelope, ts: Date): Event[] {
  if (!env.message) return [];
  const msg = env.message as Message;
  if (!msg.content) {
    return msg.role === "assistant" && msg.stop_reason === "end_turn"
      ? [makeEvent({ kind: EventKind.TurnEnd, time: ts, role: "assistant" })]
      : [];
  }

  if (typeof msg.content === "string") {
    const kind =
      msg.role === "assistant" ? EventKind.AssistantText : EventKind.UserMessage;
    return [
      makeEvent({ kind, time: ts, text: msg.content, role: msg.role }),
      ...(msg.role === "assistant" && msg.stop_reason === "end_turn"
        ? [makeEvent({ kind: EventKind.TurnEnd, time: ts, role: "assistant" })]
        : []),
    ];
  }

  if (!Array.isArray(msg.content)) return [];
  const blocks = msg.content as ContentBlock[];

  const events: Event[] = [];
  for (const b of blocks) {
    switch (b.type) {
      case "text": {
        const kind =
          msg.role === "user" ? EventKind.UserMessage : EventKind.AssistantText;
        events.push(
          makeEvent({ kind, time: ts, text: b.text ?? "", role: msg.role }),
        );
        break;
      }
      case "thinking":
        if (b.thinking) {
          events.push(
            makeEvent({
              kind: EventKind.AssistantThinking,
              time: ts,
              text: b.thinking,
              role: "assistant",
            }),
          );
        }
        break;
      case "tool_use": {
        const inp = b.input ? JSON.stringify(b.input) : "{}";
        events.push(
          makeEvent({
            kind: EventKind.ToolCall,
            time: ts,
            text: `${b.name}(${truncate(inp, 80)})`,
            toolName: b.name ?? "",
            toolID: b.id ?? "",
            toolInput: inp,
            role: "assistant",
          }),
        );
        break;
      }
      case "tool_result": {
        const text = extractToolResultText(b);
        events.push(
          makeEvent({
            kind: EventKind.ToolResult,
            time: ts,
            text: truncate(text, 200),
            toolID: b.tool_use_id ?? "",
            role: "user",
          }),
        );
        break;
      }
    }
  }
  if (msg.role === "assistant" && msg.stop_reason === "end_turn") {
    events.push(makeEvent({ kind: EventKind.TurnEnd, time: ts, role: "assistant" }));
  }
  return events;
}

export function parseLineEvents(line: string): Event[] {
  let env: Envelope;
  try {
    env = JSON.parse(line);
  } catch {
    return [];
  }

  const ts = env.timestamp ? new Date(env.timestamp) : new Date();

  switch (env.type) {
    case "user":
      return parseMessageLineEvents(env, ts);
    case "system":
      if (env.content) {
        return [
          makeEvent({
            kind: EventKind.System,
            time: ts,
            text: env.content,
            role: "system",
          }),
        ];
      }
      return [];
    case "progress": {
      const ev = parseProgress(env, ts);
      return ev ? [ev] : [];
    }
    default:
      if (env.message) {
        return parseMessageLineEvents(env, ts);
      }
      return [];
  }
}

async function readAppended(
  path: string,
  offset: number,
): Promise<{ data: string; offset: number }> {
  try {
    const fi = await stat(path);
    if (fi.size < offset) offset = 0;
    if (fi.size === offset) return { data: "", offset };

    const f = await open(path, "r");
    try {
      const buf = Buffer.alloc(fi.size - offset);
      await f.read(buf, 0, buf.length, offset);
      return { data: buf.toString(), offset: fi.size };
    } finally {
      await f.close();
    }
  } catch {
    return { data: "", offset };
  }
}

export class Tailer {
  private path: string;
  private abortController: AbortController;
  private eventQueue: Event[] = [];
  private resolvers: ((value: IteratorResult<Event>) => void)[] = [];
  private done = false;

  private startOffset: number;

  constructor(path: string, startOffset = 0) {
    this.path = path;
    this.startOffset = startOffset;
    this.abortController = new AbortController();
    this.run();
  }

  close(): void {
    this.abortController.abort();
  }

  async *events(): AsyncGenerator<Event> {
    while (true) {
      if (this.eventQueue.length > 0) {
        yield this.eventQueue.shift()!;
        continue;
      }
      if (this.done) return;
      const result = await new Promise<IteratorResult<Event>>((resolve) => {
        this.resolvers.push(resolve);
      });
      if (result.done) return;
      yield result.value;
    }
  }

  private pushEvent(ev: Event): void {
    const resolver = this.resolvers.shift();
    if (resolver) {
      resolver({ value: ev, done: false });
    } else {
      this.eventQueue.push(ev);
    }
  }

  private finish(): void {
    this.done = true;
    for (const resolver of this.resolvers) {
      resolver({ value: undefined, done: true } as IteratorResult<Event>);
    }
    this.resolvers = [];
  }

  private emitCompleteLines(pending: string): string {
    while (true) {
      const idx = pending.indexOf("\n");
      if (idx < 0) return pending;

      const line = pending.slice(0, idx).replace(/\r$/, "");
      pending = pending.slice(idx + 1);
      if (!line) continue;

      for (const ev of parseLineEvents(line)) {
        this.pushEvent(ev);
      }
    }
  }

  private async run(): Promise<void> {
    const signal = this.abortController.signal;

    while (!signal.aborted) {
      try {
        await stat(this.path);
        break;
      } catch {
        await sleep(250);
        if (signal.aborted) {
          this.finish();
          return;
        }
      }
    }

    let offset = this.startOffset;
    let pending = "";

    while (!signal.aborted) {
      const { data, offset: newOffset } = await readAppended(this.path, offset);
      offset = newOffset;
      if (data) {
        pending += data;
        pending = this.emitCompleteLines(pending);
      }

      await sleep(200);
    }

    this.finish();
  }
}
